src/org/mozilla/javascript/EqualObjectGraphs.java

```diff
@@ -63,6 +63,21 @@ boolean equalGraphs(Object o1, Object o2) {
             return true;
         } else if (o1 == null || o2 == null) {
             return false;
+            // String (and ConsStrings), Booleans, and Doubles are considered
+            // JavaScript primitive values and are thus compared by value and
+            // with no regard to their object identity.
+        } else if (o1 instanceof String) {
+            if (o2 instanceof ConsString) {
+                return o1.equals(o2.toString());
+            }
+            return o1.equals(o2);
+        } else if (o1 instanceof ConsString) {
+            if (o2 instanceof String || o2 instanceof ConsString) {
+                return o1.toString().equals(o2.toString());
+            }
+            return false;
+        } else if (o1 instanceof Boolean || o1 instanceof Double) {
```

Comment:
I think internally any `java.lang.Number` can represent a javascript primitive number, with the exception of BigInteger, which is a javascript bigint.

Reply:
Are you sure about that? Even otherwise integer literals will end up being represented as doubles in Rhino, e.g.
```java
import org.mozilla.javascript.Context;

public class TestNumbers {
    public static void main(String[] args) {
        try (var cx = Context.enter()) {
            cx.setOptimizationLevel(9);
            var scope = cx.initStandardObjects();
            var result = cx.evaluateString(scope, "1", "", 1, null);
            System.out.println(result.getClass().getName());
        }
    }
}
```
prints:
```
java.lang.Double
```
(Notice I used the most aggressive optimization level.) I could add additional types that we _certainly_ know the engine uses internally, but I wouldn't blanket treat all `Number` subclasses especially since there can be user-defined subclasses too.